File: python/multidimensional/unique_paths_II.py
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstaclesGrid: List[List[int]]) -> int:
        if len(obstaclesGrid) == 0 or len(obstaclesGrid) == 0:
            return 0


        rows = len(obstaclesGrid)
        cols = len(obstaclesGrid[0])

        if obstaclesGrid[0][0] == 1:
            return 0

        obstaclesGrid[0][0] = 1

        for j in range(1, cols):
            if obstaclesGrid[0][j] == 1:
                obstaclesGrid[0][j] = obstaclesGrid[0][j-1]



        for i in range(1, rows):
            if obstaclesGrid[i][0] == 1:
                obstaclesGrid[i][0] = 0
            else:
                obstaclesGrid[i][0] = obstaclesGrid[i-1][0]



        for i in range(1, rows):
            for j in range(1, cols):
                if obstaclesGrid[i][j] == 1:
                    obstaclesGrid[i][j] = 0
                else:
                    obstaclesGrid[i][j] = obstaclesGrid[i-1][j] + obstaclesGrid[i ][j-1]


        return obstaclesGrid[rows-1][cols-1]

File: golang/multidimensional/unique_paths_II.go
package main

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	if len(obstacleGrid) == 0 || len(obstacleGrid) == 0 {
		return 0
	}

	rows := len(obstacleGrid)
	cols := len(obstacleGrid[0])

	if obstacleGrid[0][0] == 1 {
		return 0
	}

	obstacleGrid[0][0] = 1

	for j := 1; j < cols; j++ {
		if obstacleGrid[0][j] == 1 {
			obstacleGrid[0][j] = 0
		} else {
			obstacleGrid[0][j] = obstacleGrid[0][j-1]
		}
	}

	for i := 1; i < rows; i++ {
		if obstacleGrid[i][0] == 1 {
			obstacleGrid[i][0] = 0
		} else {
			obstacleGrid[i][0] = obstacleGrid[i-1][0]
		}
	}

	for i := 1; i < rows; i++ {
		for j := 1; j < cols; j++ {
			if obstacleGrid[i][j] == 1 {
				obstacleGrid[i][j] = 0
			} else {
				obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]
			}
		}
	}

	return obstacleGrid[rows-1][cols-1]
}
